### ISSUE TITLE
Corrupt download packages for NCCL 2.5.4

### ISSUE THREAD

**Issue by cliffwoolley** [NVIDIA/nccl#269]
Transferring in from https://github.com/NVIDIA/nvidia-docker/issues/1141 .  Assigning myself.

**Comment by cliffwoolley** [comment]
Quoting the original issue:
```
[91mE: Failed to fetch https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64/./libnccl-dev_2.5.4-1+cuda10.1_amd64.deb  File has unexpected size (20038080 != 20039408). Mirror sync in progress? [IP: 192.229.211.70 443]
   Hashes of expected file:
    - SHA512:03699ded9985356a392a65f9273dd441ccded2c3a4c9bb58d212d1be7fbc30d8020a6e13c4b0261b7e3fdf3516614c1d851507a738acbf3921084270794e0f3a
    - SHA256:754b1ef020c1fb51bd9df6f1b096dd84844eca2c4916daef61731a4a951d56b4
    - SHA1:021228717e4231c78cf52bbd0c0d787018391837 [weak]
    - MD5Sum:ce4e793a440410dfa599578674a9141d [weak]
    - Filesize:20039408 [weak]
E: Failed to fetch https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64/./libnccl2_2.5.4-1+cuda10.1_amd64.deb  File has unexpected size (22972696 != 22974024). Mirror sync in progress? [IP: 192.229.211.70 443]
   Hashes of expected file:
    - SHA512:a1488e248b71de0113888bad78091133847ddd3b4b895acadc1c277dbe7110e37347e83bf5afcea6b7a840dfcaa94c24f5c25b00cac986d6bd154463b74fd50a
    - SHA256:d97322b0a4a59a5fcce23cccd5e9aad1c88f79fb9b3039923dc8b9f9dc2e6607
    - SHA1:eb8471bbd0123118903c9db472f3434acc6622e5 [weak]
    - MD5Sum:e4e650ae3e52b3b35fe34ff1982e0b42 [weak]
    - Filesize:22974024 [weak]
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
```
/cc @tgaddair

**Comment by cliffwoolley** [comment]
I confirm the md5sum mismatches are limited to NCCL 2.5.4 packages (though not just Ubuntu 18.04).  NCCL 2.5.6 is okay, as are other versions.

@tgaddair - you could switch your Horovod CI testing from 2.4.8 to 2.5.6 if you'd like, skipping over 2.5.4.  We'll get the 2.5.4 packages fixed as soon as possible, naturally, but 2.5.6 has several important bug fixes anyway.

**Comment by tgaddair** [comment]
Thanks for looking into this @cliffwoolley, we'll go ahead and move the tests to NCCL 2.5.6.

**Comment by tgaddair** [comment]
Hey @cliffwoolley, seems there's no CUDA 10.1 package for NCCL 2.5.6 available in apt, only CUDA 10.2.

**Comment by cliffwoolley** [comment]
@tgaddair That's true.   However, you can use the CUDA 10.0 build of NCCL in your otherwise-10.1-built application, and this works fine.

**Comment by tgaddair** [comment]
Hey @cliffwoolley, is that also true when using `libnccl-dev`?  I tried running with the following setup: 

```
libnccl-2=2.5.6-1+cuda10.0
libnccl-dev=2.5.6-1+cuda10.0
libcudnn7=7.6.5.32-1+cuda10.1
FROM nvidia/cuda:10.1-devel-ubuntu18.04
```

But it seems that both TensorFlow and PyTorch built for CUDA 10.1 are running into issues:

```
6f861b3d3942:15:90 [1] enqueue.cc:195 NCCL WARN Cuda failure 'invalid device function'
--
  | 6f861b3d3942:15:90 [1] NCCL INFO enqueue.cc:460 -> 1
  |  
  | 6f861b3d3942:14:91 [0] enqueue.cc:195 NCCL WARN Cuda failure 'invalid device function'
  | 6f861b3d3942:14:91 [0] NCCL INFO enqueue.cc:460 -> 1
  | terminate called after throwing an instance of 'c10::Error'
  | terminate called after throwing an instance of 'c10::Error'
  | what():    what():  CUDA error: invalid device function (launch_kernel at /pytorch/aten/src/ATen/native/cuda/Loops.cuh:103)
  | frame #0: c10::Error::Error(c10::SourceLocation, std::string const&) + 0x33 (0x7f9207d87403 in /usr/local/lib/python3.6/dist-packages/torch/lib/libc10.so)


6f861b3d3942:15:90 [1] enqueue.cc:195 NCCL WARN Cuda failure 'invalid device function'
--
  | 6f861b3d3942:15:90 [1] NCCL INFO enqueue.cc:460 -> 1
  |  
  | 6f861b3d3942:14:91 [0] enqueue.cc:195 NCCL WARN Cuda failure 'invalid device function'
  | 6f861b3d3942:14:91 [0] NCCL INFO enqueue.cc:460 -> 1
  | terminate called after throwing an instance of 'c10::Error'
  | terminate called after throwing an instance of 'c10::Error'
  | what():    what():  CUDA error: invalid device function (launch_kernel at /pytorch/aten/src/ATen/native/cuda/Loops.cuh:103)
  | frame #0: c10::Error::Error(c10::SourceLocation, std::string const&) + 0x33 (0x7f9207d87403 in /usr/local/lib/python3.6/dist-packages/torch/lib/libc10.so)
```

**Comment by cliffwoolley** [comment]
@tgaddair It should be, at least for shared libraries.  If you're using static linking (which at least PyTorch might do be default), then that could be the problem.   Sorry for the hassle here.

**Comment by tgaddair** [comment]
No worries @cliffwoolley, is it possible to release a separate 10.1 package for 2.5.6?  Seems that might simplify some things.

**Comment by cliffwoolley** [comment]
I'll have to leave it to @spotluri to reply about release plans.  Thanks!

**Comment by tgaddair** [comment]
Hey @cliffwoolley, one more thing came up that may be related to this issue.  In addition to rolling back to NCCL 2.4.8, I also needed to pin our builds to a specific digest of the `nvidia/cuda:10.1-devel-ubuntu18.04` image: `sha256:68efc9bbe07715c54ff30850aeb2e6f0d0b692af3c8dd40f13c0b179bfc0bc15`.  It seems that using newer versions of the image result in what appear to be out of memory failures with TensorFlow (see below) and MXNet:

```
2019-12-01 01:54:30.109738: E tensorflow/stream_executor/cuda/cuda_blas.cc:238] failed to create cublas handle: CUBLAS_STATUS_NOT_INITIALIZED
```

Here's a link to failing builds: https://buildkite.com/horovod/horovod/builds/1479#aa241f6c-b6a8-4948-948a-f496f8bc0361

Let me know if you think this issue is better handled by the nvidia-docker project or somewhere else.  Thanks.

**Comment by spotluri** [comment]
> No worries @cliffwoolley, is it possible to release a separate 10.1 package for 2.5.6? Seems that might simplify some things.

@tgaddair, we are currently QA'ing 10.1 packages for 2.5.6. We expect to post them early next week.

**Comment by tgaddair** [comment]
Thanks for the update @spotluri!  Next week sounds good.

**Comment by cliffwoolley** [comment]
> In addition to rolling back to NCCL 2.4.8, I also needed to pin our builds to a specific digest of the `nvidia/cuda:10.1-devel-ubuntu18.04` image: `sha256:68efc9bbe07715c54ff30850aeb2e6f0d0b692af3c8dd40f13c0b179bfc0bc15`. It seems that using newer versions of the image result in what appear to be out of memory failures with TensorFlow (see below) and MXNet:
> 
> ```
> 2019-12-01 01:54:30.109738: E tensorflow/stream_executor/cuda/cuda_blas.cc:238] failed to create cublas handle: CUBLAS_STATUS_NOT_INITIALIZED
> ```
> 
> Here's a link to failing builds: https://buildkite.com/horovod/horovod/builds/1479#aa241f6c-b6a8-4948-948a-f496f8bc0361
> 
> Let me know if you think this issue is better handled by the nvidia-docker project or somewhere else. Thanks.

@tgaddair This part is probably due to https://github.com/NVIDIA/nvidia-docker/issues/1143#issuecomment-562210002 .

**Comment by tgaddair** [comment]
Seems NCCL 2.5.6 is now available for CUDA 10.1.  Thanks @spotluri and @cliffwoolley.  Feel free to close.

**Comment by sjeaugey** [comment]
Thanks. Closing.
